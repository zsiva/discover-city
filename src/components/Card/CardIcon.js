//@flow
import React, { Component } from 'react';
import { Icon } from 'semantic-ui-react';

import './style.css';

export type CardIconPropType = {
  icon: string,
  flipped: boolean,
  handleClick: () => any,
};

export type CardIconStateType = {
  flipped: boolean,
};

class CardIcon extends Component<CardIconPropType, CardIconStateType> {
  constructor(props: CardIconPropType) {
    super(props);
    this.state = { flipped: false };
    (this: any).handleClick = this.handleClick.bind(this);
  }

  handleClick() {
    this.setState({ ...this.state, flipped: !this.state.flipped });
  }

  render() {
    return (
      <div className="card">
        <div
          className={
            this.state.flipped || this.props.flipped ? 'cardContent flipped' : ' cardContent'
          }
          onMouseDown={this.props.handleClick}
        >
          <div className="front" />
          <div className="back icon">
            <Icon size="huge" color="green" name={this.props.icon} />
          </div>
        </div>
      </div>
    );
  }
}

export default CardIcon;
