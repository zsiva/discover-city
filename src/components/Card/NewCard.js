import React, { Component } from 'react';
import { Icon } from 'semantic-ui-react';

import './style.css';

export default class NewCard extends Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(e) {
    if (!this.props.flipped) {
      this.props.checkMatch(this.props.value, this.props.id);
    }
  }

  render() {
    return (
      <div className="card">
        <div
          className={this.props.flipped ? 'cardContent flipped' : ' cardContent'}
          onMouseDown={this.handleClick}
        >
          <div className="front" />
          <div className="back icon">
            <Icon size="huge" color="green" name={this.props.value} />
          </div>
        </div>
      </div>
    );
  }
}
