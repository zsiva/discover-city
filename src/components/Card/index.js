//@flow
import React, { Component } from 'react';
import './style.css';

export type CardPropType = {
  label: string,
  img: string,
};

export type CardStateType = {
  classes: string,
};

class Card extends Component<CardPropType, CardStateType> {
  constructor(props: CardPropType) {
    super(props);
    this.state = { classes: '' };
    (this: any).handleClick = this.handleClick.bind(this);
  }

  componentWillReceiveProps(nextProps: CardPropType) {
    console.log(nextProps, this.props);
  }

  handleClick() {
    this.setState({ ...this.state, classes: 'flipped' });
  }

  render() {
    const { img, label } = this.props;
    return (
      <div className="cityCol">
        <div className={['cityImage', this.state.classes].join(' ')} onMouseDown={this.handleClick}>
          <div className="front" />
          <div className="back" style={{ backgroundImage: 'url(' + img + ')' }}>
            <span className="cityLabel">{label}</span>
          </div>
        </div>
      </div>
    );
  }
}

export default Card;
