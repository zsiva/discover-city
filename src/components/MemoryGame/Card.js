//@flow
import React, { Component } from 'react';
import { Icon } from 'semantic-ui-react';

import './style.css';

export type CardProptype = {
  flipped: boolean,
  matched: boolean,
  value: string,
  id: number,
  checkMatch: (value: string, id: number) => void,
};

export default class Card extends Component<CardProptype> {
  constructor(props: CardProptype) {
    super(props);
    (this: any).handleClick = this.handleClick.bind(this);
  }

  handleClick(e: SyntheticEvent<HTMLButtonElement>) {
    if (!this.props.flipped) {
      this.props.checkMatch(this.props.value, this.props.id);
    }
  }

  render() {
    return (
      <div className="cardContainer" onMouseDown={this.handleClick}>
        <div className={this.props.flipped ? ' flipped card' : ' card'}>
          <div className="front" />
          <div className="back">
            <Icon size="huge" color="green" name={this.props.value} />
          </div>
        </div>
      </div>
    );
  }
}
