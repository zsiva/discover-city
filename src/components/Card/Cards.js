import React, { Component } from 'react';
import { Grid } from 'semantic-ui-react';
import CardIcon from './CardIcon';
import { shuffleArray } from '../../utils/operations';

const cardIcons = ['birthday', 'alarm', 'bug', 'paint brush', 'unhide', 'pin'];

const cardList = shuffleArray([...cardIcons, ...cardIcons]);

export type CardsStateType = {
  cards: string[],
  previousAnswer: string,
  flipped: boolean,
};

export default class Cards extends Component {
  constructor(props) {
    super(props);
    this.state = {
      previousAnswer: '',
      flipped: false,
    };
    (this: any).handleClick = this.handleClick.bind(this);
  }

  handleClick(iconName: string) {
    console.log(iconName, this.state.previousAnswer);
    if (this.state.previousAnswer === iconName) {
      this.setState({ previousAnswer: '', flipped: true });
    } else {
      this.setState({ previousAnswer: iconName });
    }
  }

  render() {
    return (
      <Grid>
        {cardList.map(card => (
          <Grid.Column mobile={8} tablet={4} computer={4} onClick={() => this.handleClick(card)}>
            <CardIcon icon={card} flipped={this.state.flipped} />
          </Grid.Column>
        ))}
        <br />
      </Grid>
    );
  }
}
