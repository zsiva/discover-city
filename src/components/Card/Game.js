import React, { Component, Fragment } from 'react';
import { Grid } from 'semantic-ui-react';
import CardIcon from './CardIcon';
import NewCard from './NewCard';
import { shuffleArray } from '../../utils/operations';

const cardIcons = ['birthday', 'alarm', 'bug', 'paint brush', 'unhide', 'pin'];

const cardList = shuffleArray([...cardIcons, ...cardIcons]);

function initiateCards() {
  return cardList.map(card => ({
    value: card,
    matched: false,
    flipped: false,
  }));
}

export default class Cards extends Component {
  constructor(props) {
    super(props);
    this.renderCards = this.renderCards.bind(this);
    this.checkMatch = this.checkMatch.bind(this);
    this.reset = this.reset.bind(this);

    this.state = {
      cards: initiateCards(),
      lastCard: null,
      locked: false,
      matches: 0,
    };
  }

  checkMatch(value, id) {
    if (this.state.locked) {
      return;
    }

    var cards = this.state.cards;
    cards[id].flipped = true;
    this.setState({ cards, locked: true });
    if (this.state.lastCard) {
      if (value === this.state.lastCard.value) {
        var matches = this.state.matches;
        cards[id].matched = true;
        cards[this.state.lastCard.id].matched = true;
        this.setState({ cards, lastCard: null, locked: false, matches: matches + 1 });
      } else {
        setTimeout(() => {
          cards[id].flipped = false;
          cards[this.state.lastCard.id].flipped = false;
          this.setState({ cards, lastCard: null, locked: false });
        }, 1000);
      }
    } else {
      this.setState({
        lastCard: { id, value },
        locked: false,
      });
    }
  }

  renderCards(cards) {
    return cards.map((card, index) => {
      return (
        <Grid.Column key={index} mobile={8} tablet={4} computer={4}>
          <NewCard
            key={index}
            value={card.value}
            id={index}
            matched={card.matched}
            flipped={card.flipped}
            checkMatch={this.checkMatch}
          />
        </Grid.Column>
      );
    });
  }

  reset() {
    this.setState({
      cards: initiateCards(),
      lastCard: null,
      locked: false,
      matches: 0,
    });
  }

  render() {
    const numPairs = this.state.cards.filter(card => card.matched).length / 2;
    return (
      <Fragment>
        <h5>Num of pairs found: {numPairs}</h5>
        <Grid>{this.renderCards(this.state.cards)}</Grid>
      </Fragment>
    );
  }
}
