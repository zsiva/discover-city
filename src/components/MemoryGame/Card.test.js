import React from 'react';
import renderer from 'react-test-renderer';

import Card from './Card';

const props = {
  id: '1',
  value: 'bug',
  matched: false,
  flipped: false,
};

const mockCallBack = jest.fn();
describe('Card', () => {
  it('should match snapshot of a card', () => {
    const tree = renderer.create(<Card {...props} checkMatch={mockCallBack} />);
    expect(tree).toMatchSnapshot();
  });
  it('should match snapshot of a flipped card', () => {
    const tree = renderer.create(<Card {...props} flipped checkMatch={mockCallBack} />);
    expect(tree).toMatchSnapshot();
  });
});
