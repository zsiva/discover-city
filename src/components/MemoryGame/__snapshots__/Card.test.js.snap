// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match snapshot of a card 1`] = `
<div
  className="cardContainer"
  onMouseDown={[Function]}
>
  <div
    className=" card"
  >
    <div
      className="front"
    />
    <div
      className="back"
    >
      <i
        aria-hidden="true"
        className="green bug huge icon"
      />
    </div>
  </div>
</div>
`;

exports[`Card should match snapshot of a flipped card 1`] = `
<div
  className="cardContainer"
  onMouseDown={[Function]}
>
  <div
    className=" flipped card"
  >
    <div
      className="front"
    />
    <div
      className="back"
    >
      <i
        aria-hidden="true"
        className="green bug huge icon"
      />
    </div>
  </div>
</div>
`;
